common/src/secrets.rs

```diff
@@ -0,0 +1,117 @@
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::BTreeMap,
+    fmt::{Debug, Display},
+};
+use zeroize::Zeroize;
+
+/// Wrapper type for secret values such as passwords or authentication keys.
+///
+/// Once wrapped, the inner value cannot leak accidentally, as both the [`Display`] and [`Debug`]
+/// implementations cover up the actual value and only show the type.
+///
+/// If you need access to the inner value, there is an [expose](`Secret::expose`) method.
+///
+/// To make sure nothing leaks after the [`Secret`] has been dropped, a custom [`Drop`]
+/// implementation will zero-out the underlying memory.
+#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Secret<T: Zeroize>(T);
+
+impl<T: Zeroize> Debug for Secret<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "[REDACTED {:?}]", std::any::type_name::<T>())
+    }
+}
+
+impl<T: Zeroize> Display for Secret<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{self:?}")
+    }
+}
+
+impl<T: Zeroize> Drop for Secret<T> {
+    fn drop(&mut self) {
+        self.0.zeroize();
+    }
+}
+
+impl<T: Zeroize> From<T> for Secret<T> {
+    fn from(value: T) -> Self {
+        Self::new(value)
+    }
+}
+
+impl<T: Zeroize> Secret<T> {
+    pub fn new(secret: T) -> Self {
+        Self(secret)
+    }
+
+    pub fn expose(&self) -> &T {
+        &self.0
+    }
+}
+
+/// Store that holds all the secrets available to a deployment
+#[derive(Deserialize, Serialize, Clone)]
+pub struct SecretStore {
+    pub(crate) secrets: BTreeMap<String, Secret<String>>,
+}
+
+impl SecretStore {
+    pub fn new(secrets: BTreeMap<String, Secret<String>>) -> Self {
+        Self { secrets }
+    }
+
+    pub fn get(&self, key: &str) -> Option<String> {
+        self.secrets
+            .get(key)
+            .map(|secret| secret.expose().to_owned())
+    }
+}
+
+#[cfg(test)]
+mod secrets_tests {
+    use super::*;
+
+    #[test]
+    fn display() {
+        let password_string = String::from("VERYSECRET");
+        let secret = Secret::new(password_string);
+        let printed = format!("{}", secret);
+        assert_eq!(printed, "[REDACTED \"alloc::string::String\"]");
+    }
+
+    #[test]
+    fn debug() {
+        let password_string = String::from("VERYSECRET");
+        let secret = Secret::new(password_string);
+        let printed = format!("{:?}", secret);
+        assert_eq!(printed, "[REDACTED \"alloc::string::String\"]");
+    }
+
+    #[test]
+    fn expose() {
+        let password_string = String::from("VERYSECRET");
+        let secret = Secret::new(password_string);
+        let printed = format!("{}", secret.expose());
+        assert_eq!(printed, "VERYSECRET");
+    }
+
+    #[test]
+    fn secret_struct() {
+        #[allow(dead_code)]
```

Comment:
Initially I didn't have tests for it as I figured it's unnecessary (since everything we _could_ test is basically already enforced by the compiler), but there was [a point](https://github.com/shuttle-hq/shuttle/pull/925#discussion_r1205028767) about having them to prevent regressions in the future. I wasn't entirely sure what to do with that, so if you think we should remove them, I will remove them again.

Reply:
Ah, yeah. The compiler was complaining about the struct being unused.